File: ros2/lifecycle/src/lifecycle_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>
#include <rclcpp_lifecycle/lifecycle_publisher.hpp>

#include <lifecycle_msgs/msg/transition.hpp>

using LifecycleReturn=rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

class LifecycleNodeExample: public rclcpp_lifecycle::LifecycleNode{
public:
// Instantiation of the node
LifecycleNodeExample(const std::string & name="lifecycle_node"):
LifecycleNode(name){

}

~LifecycleNodeExample(){}

private:
// Function is called when transitioning to Configured State
LifecycleReturn on_configure(const rclcpp_lifecycle::State & state){
  RCLCPP_INFO(get_logger(),"Node '%s' is in state '%s'. Transitioning to 'configure'", get_name(), state.label().c_str());

  // If all is good, transition is successful
  return LifecycleReturn::SUCCESS;
}

// Function called when transtitionig to Active State
LifecycleReturn on_activate(const rclcpp_lifecycle::State & state){
  RCLCPP_INFO(get_logger(),"Node '%s' is in state '%s'. Transitioning to 'activate'", get_name(), state.label().c_str());

  return LifecycleReturn::SUCCESS;
}

LifecycleReturn on_deactivate(const rclcpp_lifecycle::State & state){
  RCLCPP_INFO(get_logger(),"Node '%s' is in state '%s'. Transitioning to 'deactivate'", get_name(), state.label().c_str());

  return LifecycleReturn::SUCCESS;
}

LifecycleReturn on_shutdown(const rclcpp_lifecycle::State & state){
  RCLCPP_INFO(get_logger(),"Node '%s' is in state '%s'. Transitioning to 'shutdown'", get_name(), state.label().c_str());

  return LifecycleReturn::SUCCESS;
}

};



int main(int argc, char **argv){
  rclcpp::init(argc, argv);
  auto lifecycle_node = std::make_shared<LifecycleNodeExample>("lifecycle_node");
  rclcpp::spin(lifecycle_node->get_node_base_interface());
  return 0;
}

File: ros2/lifecycle_py/lifecycle_py/lifecycle_node.py
import rclpy
from rclpy.lifecycle import State, TransitionCallbackReturn, Node

class LifecycleNodeExample(Node):
  def __init__(self, node_name,**kwargs) -> None:
    super().__init__(node_name,**kwargs)

  def on_configure(self, state: State) -> TransitionCallbackReturn:
    self.get_logger().info(f"Node '{self.get_name()}' is in state '{state.label}'. Transitioning to 'configure'")
    return TransitionCallbackReturn.SUCCESS

  def on_activate(self, state: State) -> TransitionCallbackReturn:
    self.get_logger().info(f"Node '{self.get_name()}' is in state '{state.label}'. Transitioning to 'activate'")
    return TransitionCallbackReturn.SUCCESS

  def on_deactivate(self, state: State) -> TransitionCallbackReturn:
    self.get_logger().info(f"Node '{self.get_name()}' is in state '{state.label}'. Transitioning to 'deactivate'")
    return TransitionCallbackReturn.SUCCESS

  def on_shutdown(self, state: State) -> TransitionCallbackReturn:
    self.get_logger().info(f"Node '{self.get_name()}' is in state '{state.label}'. Transitioning to 'shutdown'")
    return TransitionCallbackReturn.SUCCESS

def main(args=None) -> None:
  rclpy.init(args=args)
  lifecycle_node = LifecycleNodeExample("lifecycle_node")
  rclpy.spin(lifecycle_node)
  rclpy.shutdown()

if __name__ == "__main__":
  main()

